File: lib/airportArea.ts
import airportAreaConfig from '@/cron/internal/shared/airport-area.json';
import type { ZoneBounds } from '@/lib/geoBounds';

interface AirportCoordinates {
  latitude: number;
  longitude: number;
}

const AIRPORT_AREA_OFFSET = {
  south: airportAreaConfig.referenceAirport.latitude - airportAreaConfig.referenceBounds.south,
  west: airportAreaConfig.referenceAirport.longitude - airportAreaConfig.referenceBounds.west,
  north: airportAreaConfig.referenceBounds.north - airportAreaConfig.referenceAirport.latitude,
  east: airportAreaConfig.referenceBounds.east - airportAreaConfig.referenceAirport.longitude,
} as const;

export function getAirportAreaBounds(airport: AirportCoordinates): ZoneBounds {
  return {
    south: airport.latitude - AIRPORT_AREA_OFFSET.south,
    west: airport.longitude - AIRPORT_AREA_OFFSET.west,
    north: airport.latitude + AIRPORT_AREA_OFFSET.north,
    east: airport.longitude + AIRPORT_AREA_OFFSET.east,
  };
}

File: cron/internal/shared/airportarea.go
package shared

import (
	_ "embed"
	"encoding/json"
	"fmt"
	"sync"
)

type Bounds struct {
	South float64 `json:"south"`
	West  float64 `json:"west"`
	North float64 `json:"north"`
	East  float64 `json:"east"`
}

type ReferenceAirport struct {
	Ident     string  `json:"ident"`
	Latitude  float64 `json:"latitude"`
	Longitude float64 `json:"longitude"`
}

type airportAreaConfig struct {
	ReferenceAirport ReferenceAirport `json:"referenceAirport"`
	ReferenceBounds  Bounds           `json:"referenceBounds"`
}

type BoundsOffset struct {
	South float64
	West  float64
	North float64
	East  float64
}

//go:embed airport-area.json
var airportAreaJSON []byte

var (
	airportAreaOnce   sync.Once
	airportAreaConfigValue airportAreaConfig
	airportAreaConfigErr   error
)

func loadAirportAreaConfig() (airportAreaConfig, error) {
	airportAreaOnce.Do(func() {
		airportAreaConfigErr = json.Unmarshal(airportAreaJSON, &airportAreaConfigValue)
		if airportAreaConfigErr != nil {
			airportAreaConfigErr = fmt.Errorf("decode airport area config: %w", airportAreaConfigErr)
		}
	})

	return airportAreaConfigValue, airportAreaConfigErr
}

func MustAirportAreaConfig() airportAreaConfig {
	cfg, err := loadAirportAreaConfig()
	if err != nil {
		panic(err)
	}
	return cfg
}

func AirportAreaOffset() BoundsOffset {
	cfg := MustAirportAreaConfig()
	return BoundsOffset{
		South: cfg.ReferenceAirport.Latitude - cfg.ReferenceBounds.South,
		West:  cfg.ReferenceAirport.Longitude - cfg.ReferenceBounds.West,
		North: cfg.ReferenceBounds.North - cfg.ReferenceAirport.Latitude,
		East:  cfg.ReferenceBounds.East - cfg.ReferenceAirport.Longitude,
	}
}

func BoundsForCoordinates(latitude, longitude float64) Bounds {
	offset := AirportAreaOffset()
	return Bounds{
		South: latitude - offset.South,
		West:  longitude - offset.West,
		North: latitude + offset.North,
		East:  longitude + offset.East,
	}
}
